experiment/simulation/js/main.js

```diff
@@ -1 +1,680 @@
+    container: workingArea,
+    maxConnections: -1,
+    endpoint: {
+        type: "Dot",
+        options: { radius: 7 },
+    },
+    dragOptions: {
+        containment: "parentEnclosed",
+        containmentPadding: 5,
+    },
+    connector: "Flowchart",
+    paintStyle: { strokeWidth: 4, stroke: "#888888" },
+    connectionsDetachable: false,
+});
+
+export const connectGate = function () {
+    jsPlumbInstance.bind("beforeDrop", function (data) {
+        const fromEndpoint = data.connection.endpoints[0];
+        const toEndpoint = data.dropEndpoint;
+
+        const start_uuid = fromEndpoint.uuid.split(":")[0];
+        const end_uuid = toEndpoint.uuid.split(":")[0];
+
+        if (fromEndpoint.elementId === toEndpoint.elementId) {
+            return false;
+        }
+
+        if (start_uuid === "input" && end_uuid === "input") {
+            return false;
+        } else if (start_uuid === "output" && end_uuid === "output") {
+            return false;
+        } else {
+            jsPlumbInstance.connect({ uuids: [fromEndpoint.uuid, toEndpoint.uuid], paintStyle:{ stroke: wireColours[num_wires], strokeWidth:4 }});
+            num_wires++;
+            num_wires = num_wires % wireColours.length;
+
+            if (start_uuid === "output") {
+                const input = gatejs.gates[fromEndpoint.elementId];
+                input.isConnected = true;
+                gatejs.gates[toEndpoint.elementId].addInput(input);
+            } else if (end_uuid === "output") {
+                const input = gatejs.gates[toEndpoint.elementId];
+                input.isConnected = true;
+                gatejs.gates[fromEndpoint.elementId].addInput(input);
+            }
+
+        }
+    });
+}
+
+export const connectDecoderDisplay = function () {
+    jsPlumbInstance.bind("beforeDrop", function (data) {
+        const fromEndpoint = data.connection.endpoints[0];
+        const toEndpoint = data.dropEndpoint;
+
+        const start_uuid = fromEndpoint.uuid.split(":")[0];
+        const end_uuid = toEndpoint.uuid.split(":")[0];
+
+        if (fromEndpoint.elementId === toEndpoint.elementId) {
+            return false;
+        }
+
+        if (start_uuid === "input" && end_uuid === "input") {
+            return false;
+        } else if (start_uuid === "output" && end_uuid === "output") {
+            return false;
+        } else {
+            jsPlumbInstance.connect({ uuids: [fromEndpoint.uuid, toEndpoint.uuid], paintStyle:{ stroke: wireColours[num_wires], strokeWidth:4 }});
+            num_wires++;
+            num_wires = num_wires % wireColours.length;
+            const start_type = fromEndpoint.elementId.split("-")[0];
+            const end_type = toEndpoint.elementId.split("-")[0];
+            if (start_type === "Decoder" && end_type === "Display") {
+                if (start_uuid === "output") {
+                    let outputEndpoint = "";
+                    if (Object.keys(toEndpoint.overlays)[0].includes("a")) {

```

Comment:
This line onwards, this file has a lot of duplicated code for each of the segment. Is there a way to organise this code in functions so the duplication can be reduced?
Again, do not waste too much time on it but if there's an easy intuitive way, please use that

Reply:
I dont this so, I have to check for each endpoint and its specific to that connection. Can do it but would need to write 2-3 extra functions of same size.